src/lib/support/CHIPSparseArray.h

```diff
@@ -0,0 +1,193 @@
+/*
+ *
+ *    Copyright (c) 2020 Project CHIP Authors
+ *    Copyright (c) 2016-2017 Nest Labs, Inc.
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#pragma once
+
+#include <stddef.h>
+
+#include <core/CHIPEncoding.h>
+#include <core/CHIPError.h>
+#include <support/Base64.h>
+#include <support/CHIPMem.h>
+#include <support/CodeUtils.h>
+#include <support/SafeInt.h>
+#include <transport/raw/MessageHeader.h>
+
+namespace chip {
+
+template <uint16_t Capacity>
+class SparseU64Array
```

Comment:
At the cost of two ints, we could separate the templated bit from the non-templated bit so that code generation does not duplicate .text entries for every capactiy. Consider this design:

```cpp
class SpareU64ArrayBase {
   public:
      SpareU64ArrayBase(uint64_t *data, int16_t capacity): mData(data), mCapacity(capacity) {}
      // ...
   private:
      uint64_t *mData;
      const int16_t mCapacity;
};

template <uint16_t kCapacity>
class SparseU64Array : SparseU64ArrayBase {
    public:
       SparseU64Array() : SparseU64ArrayBase(mBuffer, kCapacity) {}
    private:
       uint64_t mBuffer[kCapacity];
};
```